### ISSUE TITLE
subtype broken for TypeVar with Union

### ISSUE THREAD

**Issue by timholy** [JuliaLang/julia#11803]
``` jl
# Good
julia> T = TypeVar(:T, Int, true)
T<:Int64

julia> T <: Int
true

# Bad
julia> T = TypeVar(:T, Union(Int,Float64), true)
T<:Union(Int64,Float64)

julia> T <: Union(Int,Float64)
false
```

**Comment by JeffBezanson** [comment]
Does this trigger any "real code" problem (other than #11015)?

I'm afraid the only real fix is in https://github.com/JuliaLang/julia/blob/master/examples/juliatypes.jl

**Comment by timholy** [comment]
I was starting to work on efficient algorithms for multiplication with SubArrays of SparseMatrixCSCs. You can write particularly-efficient algorithms if the slices are along columns, so I wanted to define this:
```jl
julia> typealias SubcolSparseMatrix{Tv,Ti,ColI<:Base.ViewIndex} SubArray{Tv,2,SparseMatrixCSC{Tv,Ti},Tuple{Colon,ColI}}
ERROR: TypeError: SubArray: in I, expected I<:Tuple{Vararg{Union{Range{Int64},Colon,Array{Int64,1},Int64}}}, got Type{Tuple{Colon,ColI<:Union{Range{Int64},Colon,Array{Int64,1},Int64}}}
```

**Comment by timholy** [comment]
I'm sure the type system rework will be the best fix, but this PR is close. The problem I've generally encountered is that often when I fix a bug in the C code, then some types "leak out" of inference in ways that they formerly hadn't. I fixed a couple of those in this PR (see the changes to inference.jl and essentials.jl), but then I ran out of time.

**Comment by JeffBezanson** [comment]
closed by #18457
